File: Source/app/src/main/java/com/amgregoire/mangafeed/Utils/NetworkService.java
package com.amgregoire.mangafeed.Utils;


import android.util.Log;

import com.amgregoire.mangafeed.Cloudflare;
import com.amgregoire.mangafeed.Cloudflare2;

import java.net.HttpCookie;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import okhttp3.Headers;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class NetworkService
{
    private static NetworkService sInstance;

    private OkHttpClient mClient;

    private NetworkService()
    {
        mClient = new OkHttpClient.Builder().readTimeout(10, TimeUnit.SECONDS)
                                            .writeTimeout(10, TimeUnit.SECONDS)
                                            .readTimeout(30, TimeUnit.SECONDS)
                                            .build();
    }

    public static Map<String, String> cookies;

    public static String defaultUserAgent = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/535.21 (KHTML, like Gecko) Chrome/19.0.1042.0 Safari/535.21";
    public static String secondUserAgent = "Mozilla/5.0 (Windows NT 6.3; WOW64)";

    public static void test()
    {
        if(cookies != null) return;

        Cloudflare2 cf = new Cloudflare2("https://www.funmanga.com/");
        cf.setUser_agent("Mozilla/5.0 (Windows NT 6.3; WOW64)");
        cf.getCookies(new Cloudflare2.cfCallback() {
            @Override
            public void onSuccess(List<HttpCookie> cookieList) {
                Log.e("cloudflare-success", cookieList.toString());
                cookies = Cloudflare2.List2Map(cookieList);
            }

            @Override
            public void onFail() {
                Log.e("cloudflare-fail", "whyyyy");
            }
        });
    }
    public static NetworkService getPermanentInstance()
    {
        test();
        if (sInstance == null)
        {
            sInstance = new NetworkService();
        }

        return sInstance;
    }

    public static NetworkService getTemporaryInstance()
    {
        test();
        return new NetworkService();
    }

    /***
     * This function returns an observable response tot he requested url.
     *
     * @param url
     * @return
     */
    public Observable<Response> getResponse(final String url)
    {
        return Observable.create((ObservableEmitter<Response> subscriber) ->
        {
            try
            {
                Request.Builder builder = new Request.Builder()
                        .url(url)
                        .header("User-Agent", secondUserAgent);

                for(String key : cookies.keySet())
                {
                    builder.addHeader("Cookie", String.format("%s=%s", key, cookies.get(key)));
                }

                Request request = builder.build();

                subscriber.onNext(mClient.newCall(request).execute());
                subscriber.onComplete();
            }
            catch (Throwable e)
            {
                subscriber.tryOnError(e);
            }
        }).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    /***
     * This function returns an observable response to the requested url, and specified headers.
     *
     * @param url
     * @param headers
     * @return
     */
    public Observable<Response> getResponseCustomHeaders(final String url, final Headers headers)
    {
        return Observable.create((ObservableEmitter<Response> subscriber) ->
        {
            try
            {
                Request.Builder builder = new Request.Builder()
                        .url(url)
                        .headers(headers);

                for(String key : cookies.keySet())
                {
                    builder.addHeader("Cookie", String.format("%s=%s", key, cookies.get(key)));
                }

                Request request = builder.build();

                subscriber.onNext(mClient.newCall(request).execute());
                subscriber.onComplete();
            }
            catch (Throwable e)
            {
                subscriber.tryOnError(e);
            }
        }).subscribeOn(Schedulers.io());
    }

    /***
     * This function returns an observable string from the specified response.
     *
     * @param response
     * @return
     */
    public static Observable<String> mapResponseToString(final Response response)
    {
        return Observable.create((ObservableEmitter<String> subscriber) ->
        {
            try
            {
                subscriber.onNext(response.body().string());
                subscriber.onComplete();
            }
            catch (Throwable e)
            {
                subscriber.tryOnError(e);
            }
        }).subscribeOn(Schedulers.io());
    }
}

File: Source/app/src/main/java/com/amgregoire/mangafeed/Utils/NetworkService.kt
package com.amgregoire.mangafeed.Utils


import com.amgregoire.mangafeed.v2.service.CloudFlareService
import com.amgregoire.mangafeed.v2.service.Logger
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.runBlocking
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.util.concurrent.TimeUnit

class NetworkService private constructor()
{

    private val mClient: OkHttpClient = OkHttpClient.Builder().readTimeout(10, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build()

    /***
     * This function returns an observable response tot he requested url.
     *
     * @param url
     * @return
     */
    fun getResponse(url: String): Observable<Response>
    {
        Logger.debug("Getting response for: $url")
        return Observable.create { subscriber: ObservableEmitter<Response> ->
            try
            {
                val builder = Request.Builder().url(url).header("User-Agent", defaultUserAgent)

                CloudFlareService().getCFCookies(url, defaultUserAgent) { cookies ->
                    for (cookie in cookies) builder.addHeader("Cookie", cookie)
                }

                val request = builder.build()

                subscriber.onNext(mClient.newCall(request).execute())
                subscriber.onComplete()
            }
            catch (e: Throwable)
            {
                subscriber.tryOnError(e)
            }
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    /***
     * This function returns an observable response to the requested url, and specified headers.
     *
     * @param url
     * @param headers
     * @return
     */
    fun getResponseCustomHeaders(url: String, headers: Headers): Observable<Response>
    {
        Logger.debug("Getting response for: $url")

        return Observable.create { subscriber: ObservableEmitter<Response> ->
            try
            {
                val builder = Request.Builder()
                        .url(url)
                        .headers(headers)

                CloudFlareService().getCFCookies(url, defaultUserAgent) { cookies ->
                    for (cookie in cookies) builder.addHeader("Cookie", cookie)
                }

                val request = builder.build()

                subscriber.onNext(mClient.newCall(request).execute())
                subscriber.onComplete()
            }
            catch (e: Throwable)
            {
                subscriber.tryOnError(e)
            }
        }.subscribeOn(Schedulers.io())
    }

    companion object
    {
        private var instance: NetworkService? = null

        val permanentInstance: NetworkService
            get() = instance ?: runBlocking {
                instance = NetworkService()
                instance!!
            }

        val temporaryInstance: NetworkService = NetworkService()

//        const val defaultUserAgent = "Mozilla/5.0 (Windows NT 6.3; WOW64)"
        const val defaultUserAgent = "Mozilla/5.0 (Linux; Android 8.0.0; SM-G960F Build/R16NW) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.84 Mobile Safari/537.36"
        /***
         * This function returns an observable string from the specified response.
         *
         * @param response
         * @return
         */
        fun mapResponseToString(response: Response): Observable<String>
        {
            return Observable.create { subscriber: ObservableEmitter<String> ->
                try
                {
                    response.body?.let { body -> subscriber.onNext(body.string()) }
                    subscriber.onComplete()
                    response.close()
                }
                catch (e: Throwable)
                {
                    subscriber.tryOnError(e)
                }
            }.subscribeOn(Schedulers.io())
        }
    }
}